src/sage/algebras/free_algebra_element.py

```diff
@@ -280,6 +280,30 @@ def _acted_upon_(self, scalar, self_on_left=False):
     # _lmul_ = _acted_upon_
     # _rmul_ = _acted_upon_
 
+    def _im_gens_(self, codomain, im_gens, base_map):
+        """
+        Apply a morphism defined by its values on the generators.
+
+        EXAMPLES::
+
+            sage: ring = algebras.Free(QQ, ['a', 'b'])
+            sage: a, b = ring.gens()
+            sage: A = matrix(QQ, 2, 2, [2, 3, 4, 1])
+            sage: B = matrix(QQ, 2, 2, [1, 7, 7, 1])
+            sage: f = ring.hom([A, B])
+            sage: f(a*b+1)
+            [24 17]
+            [11 30]
+        """
+        n = self.parent().ngens()
+        if n == 0:
+            return codomain.coerce(self)
```

Comment:
I am not sure we should coerce `self` but the underlying coefficient (i.e., we coerce the base ring element)